zcash_primitives/src/note_encryption.rs

```diff
@@ -335,35 +343,39 @@ impl SaplingNoteEncryption {
     }
 }
 
-fn parse_note_plaintext_without_memo(
+fn parse_note_plaintext_without_memo<P: consensus::Parameters>(
+    height: u32,
     ivk: &Fs,
     cmu: &Fr,
     plaintext: &[u8],
 ) -> Option<(Note<Bls12>, PaymentAddress<Bls12>)> {
     // Check note plaintext version
-    match plaintext[0] {
-        0x01 => (),
-        _ => return None,
+    if !plaintext_version_is_valid::<P>(height, plaintext[0]) {
+        return None;
     }
 
     let mut d = [0u8; 11];
     d.copy_from_slice(&plaintext[1..12]);
 
     let v = (&plaintext[12..20]).read_u64::<LittleEndian>().ok()?;
 
-    let rcm = Fs::from_repr(FsRepr(
-        plaintext[20..COMPACT_NOTE_SIZE]
-            .try_into()
-            .expect("slice is the correct length"),
-    ))?;
+    let mut r = [0u8; 32];
+    r.copy_from_slice(&plaintext[20..COMPACT_NOTE_SIZE]);
+
+    let rseed = if plaintext[0] == 0x01 {
+        let rcm = Fs::from_repr(FsRepr(r.try_into().expect("slice is the correct length")))?;
```

Comment:
The `TryInto` usage is for converting the "unknown-length" slice (which we actually know is 32 bytes) into a 32-byte array. `FsRepr` then wraps that array (which previously was implicitly specifying the type that `TryInto` was converting to).
```suggestion
    let r: [u8; 32] = plaintext[20..COMPACT_NOTE_SIZE].try_into().expect("slice is the correct length");

    let rseed = if plaintext[0] == 0x01 {
        let rcm = Fs::from_repr(FsRepr(r))?;
```